sample-apps/go-sample-app/collection/request_metrics.go

```diff
@@ -0,0 +1,110 @@
+type requestBasedMetricCollector struct {
+	totalBytesSent syncint64.Counter
+	totalRequests  asyncint64.Counter
+	latencyTime    syncint64.Histogram
+	context        context.Context
+	n              int64
+}
+
+// AddApiRequest adds 1 to the rqmc counter
+func (rqmc *requestBasedMetricCollector) AddApiRequest() {
+	atomic.AddInt64(&rqmc.n, 1)
+}
+
+// GetApiRequest gets the rqmc counter
+func (rqmc *requestBasedMetricCollector) GetApiRequest() int {
+	return int(atomic.LoadInt64(&rqmc.n))
+}
+
+// NewRequestBasedMetricCollector returns a new type struct that holds and registers the 3 request based metric instruments used in the Go-Sample-App;
+// TotalBytesSent, TotalRequests, LatencyTime
+func NewRequestBasedMetricCollector(ctx context.Context) requestBasedMetricCollector {
+	rqmc := requestBasedMetricCollector{context: ctx}
+	rqmc.registerTotalBytesSent()
+	rqmc.registerTotalRequests()
+	rqmc.registerLatencyTime()
+	return rqmc
+}
+
+// registerTotalBytesSent registers a Synchronous counter called TotalBytesSent.
+func (rqmc *requestBasedMetricCollector) registerTotalBytesSent() {
+	totalBytesSent, err := meter.SyncInt64().Counter(
+		"Total Bytes Sent",
+		instrument.WithDescription("Keeps a sum of the total amount of bytes sent while the application is alive"),
+		instrument.WithUnit("mb"),
+	)
+	if err != nil {
+		fmt.Println(err)
+	}
+	rqmc.totalBytesSent = totalBytesSent
+}
+
+// registerTotalRequests registers an Asynchronous counter called TotalApiRequests.
+func (rqmc *requestBasedMetricCollector) registerTotalRequests() {
+	totalRequests, err := meter.AsyncInt64().Counter(
+		"Total API Requests",
+		instrument.WithDescription("Increments by one every time a sampleapp endpoint is used"),
+		instrument.WithUnit("1"),
+	)
+	if err != nil {
+		fmt.Println(err)
+	}
+	rqmc.totalRequests = totalRequests
+}
+
+// registerLatencyTime registers a Synchronous histogram called LatencyTime.
+func (rqmc *requestBasedMetricCollector) registerLatencyTime() {
+	latencyTime, err := meter.SyncInt64().Histogram(
+		"Latency Time",
+		instrument.WithDescription("Measures latency time"),
+		instrument.WithUnit("ms"),
+	)
+	if err != nil {
+		fmt.Println(err)
+	}
+	rqmc.latencyTime = latencyTime
+}
+
+// StartTotalRequestCallBack starts the callback for the TotalApiRequests.
+func (rqmc *requestBasedMetricCollector) StartTotalRequestCallback() {
+	if err := meter.RegisterCallback(
+		[]instrument.Asynchronous{
+			rqmc.totalRequests,
+		},
+		// SDK periodically calls this function to collect data.
+		func(ctx context.Context) {
+			rqmc.totalRequests.Observe(ctx, int64(rqmc.GetApiRequest()))
+			fmt.Println("Total requests observed")
+		},
+	); err != nil {
+		panic(err)
+	}
+}
+
+// UpdateTotalBytesSent updates TotalBytesSent with a value between 0 and 1024
+func (rqmc *requestBasedMetricCollector) UpdateTotalBytesSent() {
+	min := 0
+	max := 1024
+	rqmc.totalBytesSent.Add(rqmc.context, int64(rand.Intn(max-min)+min))
```

Comment:
```suggestion
func (rqmc *requestBasedMetricCollector) UpdateTotalBytesSent(ctx context.Context) {
	min := 0
	max := 1024
	rqmc.totalBytesSent.Add(ctx, int64(rand.Intn(max-min)+min))
```

Any function or method that requires a context should accept it as its first parameter.  It is very rare that storing a context in a struct instance is actually the correct thing to do.

Reply:
Resolved